### ISSUE TITLE
Some spans being rejected with 400 BAD Request

### ISSUE THREAD

**Issue by Stono** [istio/istio#24177]
**Bug description**
It's some what of a follow on from https://github.com/istio/istio/issues/22968, but we're still seeing some spans get rejected with a 400 Bad Request.

Example span:
```
[
  {
    "id": "f964b08560a3e1bb",
    "tags": {
      "upstream_cluster": "inbound|80|http-web|app.sauron-web-mobile.svc.cluster.local",
      "guid:x-request-id": "59a8d4504aa2ce6f-LHR",
      "http.protocol": "HTTP/2",
      "component": "proxy",
      "http.url": "https://app.sauron-web-mobile/json/pagedata?pageId=fullPageAdvert&channel=cars&advertId=202005209430161",
      "downstream_cluster": "-",
      "user_agent": "Mozilla/5.0 (Linux; Android 9; SAMSUNG SM-A202F) AppleWebKit/537.36 (KHTML, like Gecko) SamsungBrowser/11.1 Chrome/75.0.3770.143 Mobile Safari/537.36",
      "response_flags": "-",
      "request_size": "0",
      "response_size": 1657,
      "node_id": "sidecar~10.194.32.39~sauron-web-mobile-5cdf855fbd-87zh5.sauron-web-mobile~sauron-web-mobile.svc.cluster.local",
      "http.status_code": "200",
      "http.method": "GET"
    },
    "parentId": "a9e206e25ede034a",
    "traceId": "176f1bd52837b3b5",
    "name": "app.sauron-web-mobile.svc.cluster.local:80/*",
    "timestamp": 1590677859913011,
    "duration": 1769,
    "kind": "CLIENT",
    "localEndpoint": {
      "port": 0,
      "serviceName": "sauron-web-mobile",
      "ipv4": "10.194.32.39"
    }
  }
]
```

And the error being:
```
t (.L.cAHTTP/1.1 400 Bad Request
Content-Type: text/plain; charset=utf-8
Vary: Origin
X-Content-Type-Options: nosniff
Date: Thu, 28 May 2020 14:57:39 GMT
Content-Length: 108

Unable to process request body: json: cannot unmarshal number into Go struct field Span.tags of type string
```

The problem field being:
```
"response_size": 1657,
```

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[x] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**
All spans to be sent to Jaeger in a format that jaeger can handle.

**Steps to reproduce the bug**
I've not found a consistent way to reproduce but we get a few per minute in production across a variety of apps.

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**
1.5.4

**How was Istio installed?**
Helm

**Environment where bug was observed (cloud vendor, OS, etc)**

**Comment by Stono** [comment]
Literally found one other mention of this on the whole internet haha - https://github.com/Kong/kong-plugin-zipkin/issues/8

Their fix was to ensure all the span fields were toString'd

**Comment by douglas-reid** [comment]
A naive look at the [code](https://github.com/envoyproxy/envoy/blob/8ae7046a1e16ba43420cf66d43dd0d69b500a6ca/source/common/tracing/http_tracer_impl.cc#L202) suggests that this should be a string:

> span.setTag(Tracing::Tags::get().ResponseSize, std::to_string(stream_info.bytesSent()));

**Comment by Stono** [comment]
Just for the sake of verbosity:

- In a single report, you get multiple spans (an array), it's quite common for just one of them to be wrong, and the rest fine.
- The only value we see sometimes coming through as a number is `response_size`, never `request_size`
- We see a variety of different values for response_size, so don't believe there to be any correlation as to the value
- There's no other correlation between services, response codes etc

**Comment by douglas-reid** [comment]
@objectiser any ideas? I was not able to replicate this in unit tests with some naive attempts and didn't see anything overtly problematic in the code.

This apparently happens 1/4000th of the time.

**Comment by Stono** [comment]
Did a slightly larger sample just to confirm, more like 1/1800....! 

```
root@istio-tracing-54f89c8554-7pkpk:/# tcpdump -A -s 0 'tcp port 9411 and (((ip[2:2] - ((ip[0]&0xf)<<2)) - ((tcp[12]&0xf0)>>2)) != 0)' | grep 'HTTP/1.1 ' > 1
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on eth0, link-type EN10MB (Ethernet), capture size 262144 bytes
^C103776 packets captured
103879 packets received by filter
0 packets dropped by kernel

root@istio-tracing-54f89c8554-7pkpk:/# cat 1 | grep 400 | wc -l
26

root@istio-tracing-54f89c8554-7pkpk:/# cat 1 | grep 202 | wc -l
48459
```

**Comment by objectiser** [comment]
@douglas-reid Seems strange. Had a look for anywhere that might set the response size as a number rather than string but can't find anything.

My only thought is that the fix for https://github.com/envoyproxy/envoy/pull/10400 is inadvertantly stripping the quotes from the `response_size` in rare cases where it has the same value as the `duration`. To avoid the scientific representation for the timestamp and duration fields, it stores them as strings and then post serialization removes the quotes. However if other fields (tags) happen to have the same value, then I guess it would also strip the quotes from them as well.

**Comment by objectiser** [comment]
Modified one of the unit tests to use a tag with the same value as the duration, and it did strip the quotes. So seems like this might be the problem.

@Stono Would it be possible to capture one example of the failure to confirm whether the value in `response_size` tag is the same as the duration or timestamp (less likely)?

**Comment by Stono** [comment]
@objectiser here's a collection of them that i shared with Doug yesterday:

```
  {
    "traceId": "7f48b987603d1d06",
    "name": "app.promoted-listing-service.svc.cluster.local:80/*",
    "timestamp": 1590678601668884,
    "duration": 37276,
    "kind": "CLIENT",
    "localEndpoint": {
      "port": 0,
      "serviceName": "promoted-listing-service",
      "ipv4": "10.194.35.23"
    },
    "id": "e52bea150371b9ec",
    "tags": {
      "response_flags": "-",
      "request_size": "0",
      "response_size": 37276,
      "node_id": "sidecar~10.194.35.23~promoted-listing-service-744686cf68-fz4gv.promoted-listing-service~promoted-listing-service.svc.cluster.local",
      "http.status_code": "200",
      "http.method": "GET",
      "upstream_cluster": "inbound|80|http-web|app.promoted-listing-service.svc.cluster.local",
      "guid:x-request-id": "59a8e66b8cd1d210-MAN",
      "http.protocol": "HTTP/2",
      "component": "proxy",
      "http.url": "https://app.promoted-listing-service.prod.k8.atcloud.io/promoted-listings?advert_classification=standard&advertising_location=at_cars&make=KIA&model=CEED&max_price=4000&distance=100&postcode=HP12+3NZ&fuel_type=Petrol&min_year_manufactured=2012&max_mileage=70000&condition=New&condition=Nearly+New&condition=Used",
      "downstream_cluster": "-",
      "user_agent": "Apache-HttpClient/4.5.12 (Java/1.8.0_102)"
    },
    "parentId": "8c00bb131c3b0ef4"
  },
```

```
  {
    "duration": 5719,
    "kind": "CLIENT",
    "localEndpoint": {
      "serviceName": "sauron-web",
      "ipv4": "10.194.58.27",
      "port": 0
    },
    "id": "32edd68fb252a8a2",
    "tags": {
      "http.status_code": "200",
      "http.method": "GET",
      "upstream_cluster": "outbound|80||app.search-one-read.svc.cluster.local",
      "guid:x-request-id": "59a8e65eeccce63c-LHR",
      "http.protocol": "HTTP/1.1",
      "component": "proxy",
      "http.url": "http://app.search-one-read/api/v1/retailer/329839",
      "downstream_cluster": "-",
      "user_agent": "Apache-HttpClient/4.5.12 (Java/11.0.7)",
      "response_flags": "-",
      "request_size": "0",
      "response_size": 1462,
      "node_id": "sidecar~10.194.58.27~sauron-web-84dd4d4ff8-5qdsh.sauron-web~sauron-web.svc.cluster.local"
    },
    "parentId": "3ee0950c6ba78c27",
    "traceId": "2045384f257d56ad",
    "name": "app.search-one-read.svc.cluster.local:80/*",
    "timestamp": 1590678599562149
  },
```

```
  {
    "timestamp": 1590678584564544,
    "duration": 26266,
    "kind": "CLIENT",
    "localEndpoint": {
      "serviceName": "sauron-web-mobile",
      "ipv4": "10.194.63.51",
      "port": 0
    },
    "id": "661730cfa7c1acf9",
    "tags": {
      "http.status_code": "200",
      "http.method": "GET",
      "upstream_cluster": "outbound|80||app.search-one-read.svc.cluster.local",
      "guid:x-request-id": "59a8e6015da406f9-LHR",
      "http.protocol": "HTTP/1.1",
      "component": "proxy",
      "http.url": "http://app.search-one-read/api/v1/advert?advert_classification=standard&advertising_location=at_cars&transmission=Automatic&min_year_manufactured=2019&model=A5&make=AUDI&max_year_manufactured=new&condition=New&condition=Nearly+New&condition=Used&include_finance_option=10000_1000_48",
      "downstream_cluster": "-",
      "user_agent": "Apache-HttpClient/4.5.12 (Java/11.0.7)",
      "response_flags": "-",
      "request_size": "0",
      "response_size": 4384,
      "node_id": "sidecar~10.194.63.51~sauron-web-mobile-5cdf855fbd-vnggr.sauron-web-mobile~sauron-web-mobile.svc.cluster.local"
    },
    "parentId": "f4dcbc5301168a13",
    "traceId": "d176e0ad808dc43a",
    "name": "app.search-one-read.svc.cluster.local:80/*"
  },
```

```
  {
    "timestamp": 1590678575725147,
    "duration": 1023,
    "kind": "CLIENT",
    "localEndpoint": {
      "serviceName": "search-one-read",
      "ipv4": "10.194.29.46",
      "port": 0
    },
    "id": "4ce983912df585e9",
    "tags": {
      "downstream_cluster": "-",
      "user_agent": "Solr[org.apache.solr.client.solrj.impl.HttpSolrClient] 1.0",
      "request_size": "253",
      "response_flags": "-",
      "response_size": 29434,
      "node_id": "sidecar~10.194.29.46~search-one-read-89ddc94bc-gc6kq.search-one-read~search-one-read.svc.cluster.local",
      "http.status_code": "200",
      "http.method": "POST",
      "upstream_cluster": "outbound|80||solr.search-solr.svc.cluster.local",
      "guid:x-request-id": "59a8e5ca18f0ee3b-CDG",
      "http.protocol": "HTTP/1.1",
      "component": "proxy",
      "http.url": "http://solr.search-solr/solr/tech-data/get"
    },
    "parentId": "e32e31a2e53979f0",
    "traceId": "987952d0cd1a0951",
    "name": "solr.search-solr.svc.cluster.local:80/*"
  },
```

```
  {
    "id": "566334ebaefd7ed3",
    "tags": {
      "response_size": 25774,
      "node_id": "sidecar~10.194.42.25~promoted-listing-service-744686cf68-blrr5.promoted-listing-service~promoted-listing-service.svc.cluster.local",
      "http.status_code": "200",
      "http.method": "GET",
      "upstream_cluster": "inbound|80|http-web|app.promoted-listing-service.svc.cluster.local",
      "guid:x-request-id": "59a8e5bf3867ce5f-LHR",
      "http.protocol": "HTTP/2",
      "component": "proxy",
      "http.url": "http://app.promoted-listing-service/promoted-listings?advert_classification=standard&advertising_location=at_cars&postcode=me138ja&distance=1500&make=SEAT&model=LEON&aggregated_trim=Cupra&transmission=Automatic&fuel_type=Petrol&condition=New&condition=Nearly+New&condition=Used&include_finan",
      "user_agent": "Apache-HttpClient/4.5.12 (Java/11.0.7)",
      "downstream_cluster": "-",
      "response_flags": "-",
      "request_size": "0"
    },
    "parentId": "595e7cd8d7dc66cd",
    "traceId": "079a1126b10fc224",
    "name": "app.promoted-listing-service.svc.cluster.local:80/*",
    "timestamp": 1590678573994797,
    "duration": 24498,
    "kind": "CLIENT",
    "localEndpoint": {
      "port": 0,
      "serviceName": "promoted-listing-service",
      "ipv4": "10.194.42.25"
    }
  }
```

**Comment by objectiser** [comment]
@Stono Yeah thanks - that confirms the issue - it is because the replacement of the values to remove the quotes occurs as it building up the json string response, and because it does "replaceAll", it ends up replacing other field values with the same numerical value.

**Comment by Stono** [comment]
Great stuff, glad we know what it is.  @douglas-reid can we please backport this to 1.5 when patched?  

Realistically it will be a while before my prod clusters move to 1.6 and this is a here and now issue for me!

**Comment by douglas-reid** [comment]
@objectiser fantastic work! I was trying values I thought may be problematic, but didn't make the connection to duration. well done.

and what a mess.

@Stono yes, I think we need to work to get this back to 1.5.

**Comment by Stono** [comment]
Hey @douglas-reid - just noticed this was merged to release-1.5, was that in 1.5.5?

**Comment by douglas-reid** [comment]
@fpesce to confirm, but I believe that 1.5.5 was security fix only and 1.5.6 will contain the good stuff.

**Comment by fpesce** [comment]
> @fpesce to confirm, but I believe that 1.5.5 was security fix only and 1.5.6 will contain the good stuff.

@Stono / @douglas-reid : confirming, 1.5.5 was a 0-day vulnerability we had to release ASAP.
The "good stuff" is built since this morning, but still under qualification (for 48 hours as usual for Istio). If you want to try it however, the artefacts are available here: https://pantheon.corp.google.com/storage/browser/istio-prerelease/prerelease/1.5.6

**Comment by Stono** [comment]
Thanks for confirming!
